src/transaction/gas_price.rs

```diff
@@ -0,0 +1,274 @@
+    /// otherwise.
+    pub fn value(&self) -> Option<U256> {
+        match self {
+            GasPrice::Value(value) => Some(*value),
+            _ => None,
+        }
+    }
+
+    /// Resolves the gas price into a value. Returns a future that resolves once
+    /// the gas price is calculated as this may require contacting the node for
+    /// gas price estimates in the case of `GasPrice::Standard` and
+    /// `GasPrice::Scaled`.
+    pub fn resolve<T: Transport>(web3: &Web3<T>, gas_price: Self) -> ResolveGasPriceFuture<T> {
+        ResolveGasPriceFuture::new(web3, gas_price)
+    }
+
+    /// Resolves the gas price into an `Option<U256>` intendend to be used by a
+    /// `TransactionRequest`. Note that `TransactionRequest`s gas price default
+    /// to the node's estimate (i.e. `GasPrice::Standard`) when omitted, so this
+    /// allows for a small optimization by foregoing a JSON RPC request.
+    pub fn resolve_for_transaction_request<T: Transport>(
+        web3: &Web3<T>,
+        gas_price: Option<Self>,
+    ) -> ResolveTransactionRequestGasPriceFuture<T> {
+        let future = match gas_price {
+            None | Some(GasPrice::Standard) => None,
+            Some(gas_price) => Some(GasPrice::resolve(web3, gas_price)),
+        };
+        future.into()
+    }
+}
+
+impl Default for GasPrice {
+    fn default() -> Self {
+        GasPrice::Standard
+    }
+}
+
+impl From<U256> for GasPrice {
+    fn from(value: U256) -> Self {
+        GasPrice::Value(value)
+    }
+}
+
+macro_rules! impl_gas_price_from_integer {
+    ($($t:ty),* $(,)?) => {
+        $(
+            impl From<$t> for GasPrice {
+                fn from(value: $t) -> Self {
+                    GasPrice::Value(value.into())
+                }
+            }
+        )*
+    };
+}
+
+impl_gas_price_from_integer! {
+    i8, i16, i32, i64, i128, isize,
+    u8, u16, u32, u64, u128, usize,
+}
+
+/// Future for resolving gas prices.
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[pin_project]
+pub struct ResolveGasPriceFuture<T: Transport> {
+    /// The state of the future.
+    #[pin]
+    state: ResolveGasPriceState<T>,
+}
+
+/// The state of the `ResolveGasPriceFuture`. This type is used to not leak
+/// internal implementation details.
+#[pin_project]
+enum ResolveGasPriceState<T: Transport> {
+    /// The gas price is known before hand.
+    Ready(U256),
+    /// The gas price estimate is being queried with the node. Optinally, the
+    /// gas price will be scaled once retrieved.
+    Estimating(#[pin] CompatCallFuture<T, U256>, Option<f64>),
+}
+
+impl<T: Transport> ResolveGasPriceFuture<T> {
+    /// Creates a new future that resolves once the gas price value is computed.
+    pub fn new(web3: &Web3<T>, gas_price: GasPrice) -> Self {
+        let state = match gas_price {
+            GasPrice::Standard => {
+                ResolveGasPriceState::Estimating(web3.eth().gas_price().compat(), None)
```

Comment:
Nit, could be 1.0?

Reply:
Then we could have rounding issues.

Reply:
Tried to improve the comments as to why `GasPrice::Standard` and `GasPrice::Scaled(1.0)` are slightly different:
https://github.com/gnosis/ethcontract-rs/pull/142/files#diff-b3a3e64ed50fef4a4fee763a076dc177R23-R27